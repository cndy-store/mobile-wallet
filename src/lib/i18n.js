import NativeLocale from 'react-native-locale';

const localeSettings = NativeLocale.constants();

export { localeSettings };
