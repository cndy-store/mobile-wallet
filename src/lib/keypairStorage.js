import * as Keychain from 'react-native-keychain';

const getItem = async _key => {
  const credentials = await Keychain.getGenericPassword();
  return Promise.resolve(credentials.password);
};

const setItem = (key, value) => {
  return Keychain.setGenericPassword(key, value);
};

const removeItem = _key => {
  return Keychain.resetGenericPassword();
};

export { getItem, setItem, removeItem };
