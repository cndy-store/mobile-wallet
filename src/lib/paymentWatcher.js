let timeout = null;
let pollingInterval = 3000;

const start = (dispatcher, loadFn, publicKey) => {
  if (timeout) {
    return;
  }

  timeout = window.setTimeout(() => {
    dispatcher(loadFn(publicKey))
      .then(() => {
        timeout = null;
        start(dispatcher, loadFn, publicKey);
      })
      .catch(error => {
        console.error(error);
      });
  }, pollingInterval);
};

const stop = () => {
  if (timeout) window.clearTimeout(timeout);
  timeout = null;
};

export { start, stop };
