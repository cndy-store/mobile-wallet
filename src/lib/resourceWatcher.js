let timeouts = {};
let pollingInterval = 3000;

const start = (name, dispatcher, loadFn, publicKey) => {
  if (timeouts[name]) {
    return;
  }

  timeouts[name] = window.setTimeout(() => {
    dispatcher(loadFn(publicKey))
      .then(() => {
        timeouts[name] = null;
        start(name, dispatcher, loadFn, publicKey);
      })
      .catch(error => {
        console.error(error);
      });
  }, pollingInterval);
};

const stop = name => {
  if (timeouts[name]) window.clearTimeout(timeouts[name]);
  timeouts[name] = null;
};

const startPaymentsWatcher = start.bind(null, 'payments');
const stopPaymentsWatcher = stop.bind(null, 'payments');

const startAcccountWatcher = start.bind(null, 'account');
const stopAccountWatcher = stop.bind(null, 'account');

export {
  startPaymentsWatcher,
  stopPaymentsWatcher,
  startAcccountWatcher,
  stopAccountWatcher
};
