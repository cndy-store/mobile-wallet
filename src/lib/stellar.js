import axios from 'axios';
import StellarSdk from '../vendor/stellar-base-0.7.7';
StellarSdk.Network.useTestNetwork();

const assetCode = 'CNDY';
const issuerPublicKey =
  'GCJKC2MI63KSQ6MLE6GBSXPDKTDAK43WR522ZYR3F34NPM7Z5UEPIZNX';

const asset = new StellarSdk.Asset(assetCode, issuerPublicKey);

const axiosInstance = axios.create({
  baseURL: 'https://horizon-testnet.stellar.org',
  timeout: 10 * 1000
});

module.exports = {
  axios: axiosInstance,
  asset,
  StellarSdk
};
