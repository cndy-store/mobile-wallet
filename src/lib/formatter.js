import { padEnd, reduceRight } from 'lodash';
import { localeSettings } from './i18n';

const delimiter = localeSettings.decimalSeparator;
const finalDelimiter = '.';
const maxPrecisionLength = 7;

const allowedChars = new RegExp(`^[0-9${delimiter}${finalDelimiter}]+`);
const delimiterSplit = new RegExp(`[${delimiter}${finalDelimiter}]+`);

const sanitizeOngoingAmountInput = input => {
  const onlyAllowedChars = input.match(allowedChars);
  if (!onlyAllowedChars) return '';
  if (onlyAllowedChars[0] !== input) return onlyAllowedChars[0];

  const [precision, ...rest] = input.split(delimiterSplit);
  if (!precision.length) return '';
  if (!rest.length) return precision;

  let scale = rest.join('');
  if (scale.length > maxPrecisionLength) {
    scale = scale.substr(0, maxPrecisionLength);
  }

  return [precision, scale].join(delimiter);
};

const ensureScale = input => {
  const [precision, scale] = input.split(delimiter);
  const paddedScale = padEnd(scale || '', 2, '0');

  return `${precision}${delimiter}${paddedScale}`;
};

const parseTransactionAmount = input => {
  const sanitized = sanitizeOngoingAmountInput(input);
  if (sanitized === '') return null;
  const withScale = ensureScale(sanitized);

  const withCorrectDelimiter = withScale.replace(delimiter, finalDelimiter);
  return withCorrectDelimiter;
};

const optimizeScale = scale => {
  const scaleNumbers = scale.split('');
  const neededNumbers = reduceRight(
    scaleNumbers,
    (acc, number) => {
      if (acc.length || number !== '0') {
        acc.push(number);
      }
      return acc;
    },
    []
  );

  return neededNumbers.reverse().join('');
};

// this method is to be used with normalized amount strings
const shortFormat = input => {
  const [precision, scale] = input.split(finalDelimiter);
  const optimizedScale = optimizeScale(scale);
  const paddedScale = padEnd(optimizedScale, 2, '0');

  return `${precision}${delimiter}${paddedScale}`;
};

export { sanitizeOngoingAmountInput, parseTransactionAmount, shortFormat };
