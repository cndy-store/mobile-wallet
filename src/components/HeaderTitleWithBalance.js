import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Platform } from 'react-native';
import { Title } from 'native-base';
import { asset } from '../lib/stellar';

const defaultText = '...';

export class HeaderTitleWithBalance extends Component {
  getBalance() {
    if (!this.props.data) return null;

    const assetToShow = this.props.data.balances.find(
      item =>
        item.asset_code === asset.getCode() &&
        item.asset_issuer === asset.getIssuer()
    );

    if (!assetToShow) return null;

    const balance = parseFloat(assetToShow.balance, 10);

    return `~${balance.toFixed(2)} ${asset.getCode()}`;
  }

  render() {
    const text = this.getBalance() || defaultText;

    return <Title>{text}</Title>;
  }
}

HeaderTitleWithBalance.defaultProps = {
  data: null
};

HeaderTitleWithBalance.propTypes = {
  data: PropTypes.shape({
    balances: PropTypes.array.isRequired
  })
};

const mapStateToProps = state => ({
  data: state.account.data
});

const mapDispatchToProps = dispatch => ({});

export default connect(mapStateToProps, mapDispatchToProps)(
  HeaderTitleWithBalance
);
