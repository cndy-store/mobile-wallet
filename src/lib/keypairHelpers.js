import { asset, StellarSdk } from '../lib/stellar';
import bip21 from 'bip21';

const { StrKey } = StellarSdk;

const scheme = asset.getCode().toLowerCase();
const secretScheme = `${scheme}secret`;

const isValidPublicKey = publicKey => StrKey.isValidEd25519PublicKey(publicKey);
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

const encodePublicKey = (publicKey, options = {}) => {
  if (!isValidPublicKey(publicKey)) return null;
  options.scheme = scheme;
  return bip21.encode(publicKey, options);
};

const decodePublicKey = url => {
  let decoded;
  try {
    decoded = bip21.decode(url, scheme);
  } catch (e) {
    return null;
  }

  const publicKey = decoded.address;
  if (!isValidPublicKey(publicKey)) return null;

  return {
    publicKey,
    options: decoded.options
  };
};

const encodeSecret = (secret, options = {}) => {
  if (!isValidSecret(secret)) return null;
  options.scheme = secretScheme;
  return bip21.encode(secret, options);
};

const decodeSecret = url => {
  let decoded;
  try {
    decoded = bip21.decode(url, secretScheme);
  } catch (e) {
    return null;
  }

  const secret = decoded.address;
  if (!isValidSecret(secret)) return null;

  return {
    secret,
    options: decoded.options
  };
};

export {
  isValidPublicKey,
  isValidSecret,
  encodePublicKey,
  decodePublicKey,
  encodeSecret,
  decodeSecret
};
