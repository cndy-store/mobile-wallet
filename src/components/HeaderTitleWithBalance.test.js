import React from 'react';
import { render } from '../__tests__/renderer';
import { asset } from '../lib/stellar';
import { HeaderTitleWithBalance } from './HeaderTitleWithBalance';

it('renders correctly when no balance is given', () => {
  const { toJSON } = render(<HeaderTitleWithBalance />);
  expect(toJSON()).toMatchSnapshot();
});

it('renders correctly when account data is present, but not for the desired asset', () => {
  const data = {
    balances: [
      {
        balance: '9999.9999000',
        asset_type: 'native'
      }
    ]
  };

  const { toJSON } = render(<HeaderTitleWithBalance data={data} />);
  expect(toJSON()).toMatchSnapshot();
});

describe('renders correctly when a balance is given', () => {
  it('for ios', () => {
    HeaderTitleWithBalance.prototype.getPlatform = jest.fn(() => 'ios');
    const data = {
      balances: [
        {
          balance: '930.0000000',
          limit: '922337203685.4775807',
          asset_type: 'credit_alphanum4',
          asset_code: asset.getCode(),
          asset_issuer: asset.getIssuer()
        },
        {
          balance: '9999.9999000',
          asset_type: 'native'
        }
      ]
    };

    const { toJSON } = render(<HeaderTitleWithBalance data={data} />);
    expect(toJSON()).toMatchSnapshot();
  });


  it('for android', () => {
    HeaderTitleWithBalance.prototype.getPlatform = jest.fn(() => 'android');
    const data = {
      balances: [
        {
          balance: '930.0000000',
          limit: '922337203685.4775807',
          asset_type: 'credit_alphanum4',
          asset_code: asset.getCode(),
          asset_issuer: asset.getIssuer()
        },
        {
          balance: '9999.9999000',
          asset_type: 'native'
        }
      ]
    };

    const { toJSON } = render(<HeaderTitleWithBalance data={data} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
