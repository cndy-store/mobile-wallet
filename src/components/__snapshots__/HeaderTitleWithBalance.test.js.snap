// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when a balance is given 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  numberOfLines={1}
  style={
    Object {
      "color": "#000",
      "fontFamily": "System",
      "fontSize": 17,
      "fontWeight": "600",
      "textAlign": "center",
    }
  }
  virtual={undefined}
>
  ~930.00 üç≠
</Text>
`;

exports[`renders correctly when account data is present, but not for the desired asset 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  numberOfLines={1}
  style={
    Object {
      "color": "#000",
      "fontFamily": "System",
      "fontSize": 17,
      "fontWeight": "600",
      "textAlign": "center",
    }
  }
  virtual={undefined}
>
  ...
</Text>
`;

exports[`renders correctly when no balance is given 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  numberOfLines={1}
  style={
    Object {
      "color": "#000",
      "fontFamily": "System",
      "fontSize": 17,
      "fontWeight": "600",
      "textAlign": "center",
    }
  }
  virtual={undefined}
>
  ...
</Text>
`;
